readme.md
# Payments

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Total Downloads][ico-downloads]][link-downloads]
[![Build Status][ico-travis]][link-travis]

This is simple and easy to use Laravel and Lumen package for using multiple payment gateways in a single place. Now you don't need to install multiple packages for multiple payment gateways.
Here we provide 2 of most widely used Payment Gateways (PayPal, Stripe) having easy to use methods such as `pay()`, `invoice()`, `refund()` etc.


## Installation

Via Composer

for both `Laravel` and `Lumen`

``` bash
$ composer require hestalabs/payments ">=1.0"
```

or you can just add it in your composer.json

```
"require": {
    "hestalabs/payments": ">=1.0"
}
```

Next, run `composer update`.


## Usage

i). For `Laravel` Add the ServiceProvider to your `config/app.php`

	a). Add in `providers` array,

	'providers' => array(
	    // ...

	    Hestalabs\Payments\PaymentsServiceProvider::class,
	);

	b). Add alias in `alias` array,

	'aliases' => array(
	    // ...

	    'Payment'   => Hestalabs\Payments\Facades\Payment::class,
	);

	c). Finally publish the package configurations by running the following command in `Terminal`

	`php artisan vendor:publish --provider="Hestalabs\Payments\PaymentsServiceProvider"`


ii). a). For `Lumen` make a directory named `config` and make a file there named `payments` and add the below code,

	<?php

	return [
		'payment_type' => env('PAYMENT_TYPE', 'paypal'),
		'stripe' => [
			'key'       => env('STRIPE_KEY'),
			'secret'    => env('STRIPE_SECRET'),
		],
		'paypal_payment' => [
			# Define your application mode here
			'mode' => env('PAYPAL_MODE', 'sandbox'),  //"sanbox" for testing and "live" for production

			# Account credentials from developer portal
			'account' => [
				'client_id'		=> env('PAYPAL_CLIENT_ID', ''),
				'client_secret' => env('PAYPAL_CLIENT_SECRET', ''),
			],

			# Connection Information
				'http' => [
				'connection_time_out' => 30,
				'retry' => 1,
			],

			# Logging Information
			'log' => [
				'log_enabled' => true,

				# When using a relative path, the log file is created
				# relative to the .php file that is the entry point
				# for this request. You can also provide an absolute
				# path here
				'file_name' => '../PayPal.log',

				# Logging level can be one of FINE, INFO, WARN or ERROR
				# Logging is most verbose in the 'FINE' level and
				# decreases as you proceed towards ERROR
				'log_level' => 'FINE',
			],
		],
	];

	b). Make a directory inside in `app` named `Support` and make a file there named `helpers.php`. Now add the following lines in your `composer.json`.

	//composer.json
	"autoload": {
        "classmap": [
            "database/seeds",
            "database/factories"
        ],
        "psr-4": {
            "App\\": "app/"
        },
        "files": [
            "app/Support/helpers.php"
        ]
    },

	//in app/Support/helpers.php

	<?php

	if ( ! function_exists('config_path')){
	    /**
	     * Get the configuration path.
	     *
	     * @param  string $path
	     * @return string
	     */
	    function config_path($path = ''){
	        return app()->basePath() . '/config' . ($path ? '/' . $path : $path);
	    }
	}


	c). Go to your `bootstrap/app.php` and uncomment line there `withEloquent()` and add the following lines,

	$app->register(Hestalabs\Payments\PaymentsServiceProvider::class);

	class_alias('Hestalabs\Payments\Facades\Payment', 'Payment');

	$app->configure('payments');

## Important instructions

a). search for `vendor/paypal/rest-api-sdk-php/lib/PayPal/Common/PayPalModel.php` at line 173 

change `else if(sizeof($v) <= 0 && is_array($v))` to `else if(is_array($v) && sizeof($v) <= 0)`

b). Add credentials in your .env

```
STRIPE_KEY='YOUR_CREDENTIALS'
STRIPE_SECRET='YOUR_CREDENTIALS'

PAYPAL_MODE='YOUR_CREDENTIALS' 			//"sanbox" for testing and "live" for production (default: sandbox)
PAYMENT_TYPE='YOUR_TYPE' 				//paypal, stripe (default: paypal)
PAYPAL_CLIENT_ID='YOUR_CREDENTIALS'
PAYPAL_CLIENT_SECRET='YOUR_CREDENTIALS'
```

## Examples

1). For PayPal (if `PAYMENT_TYPE` == paypal in .env)
	
	a). For payment through PayPal

	//for shipping details 						(Optional)
	$shipping = \Payment::shipping();
	$shipping->setLine1("3909 Witmer Road")
	->setLine2("Niagara Falls")
	->setCity("Niagara Falls")
	->setState("NY")
	->setPostalCode("14305")
	->setCountryCode("US")
	->setPhone("716-298-1822")
	->setRecipientName("Jhone");
		
	//if making payment via credit_card (Required if card is set otherwise not required)
	$card = \Payment::card();
	$card->setType('visa')						
	->setNumber(4032038634486363)	//required		
	->setExpireMonth(11)	//required	
	->setExpireYear(2023)	//required		
	->setCvv(123)			//required		
	->setFirstName('Abhishek')					
	->setLastName('Rana');						
		
	//if user wants to add some extra details 	(Optional)
	$details = \Payment::details();
	$details//>setShippingCharge(1.0)				
	->setShippingTax(1.0)	(sub total must be a sumation of all taxes and price)	
	->setSubtotal(2.0);

	//a). making payment (paypal checkout)
	$pay = \Payment::setTax(0.2)									
	->setPrice(1)	//required                          		
	->setQuantity(1)	//required							
	->setSubTotal(1)															
	->setTotal(1.0)	//required                          	
	->setCurrency('USD')	(Default USD, otherwise can be provided e.g. USD, AUD etc.)						
	->setDescription('New description')			
	->setItemName('New Items for shopping')	//required		
	->setShippingDetails($shipping)	//required only if shipping details are set as above, otherwise optional		
	->setReturnUrl(url('/payment?success=true'))	//required if card is not set (must be a full fletched URl)
	->setCancelUrl(url('/payment?success=false'))	//required if card is not set (must be a full fletched URl)
	->setExtraParam([		
		'url' => true	//required if you want only redirect URL for PayaPal (useful in cases when you are making API or something). if it is not set then for payment it will redirect to paypal payment screen
	])
	->setIntent('sale')	(Optional)(Default : sale)(e.g. sale, order, authorize)					
	->setDetails($details)						
	->pay();	//required

	//b). making payment through card, then set card
	$pay = \Payment::setTax(0.2)									
	->setPrice(1)	//required                          		
	->setQuantity(1)	//optional							
	->setSubTotal(1)	//optional
	->setAmount(1)		//required           	
	->setCurrency('USD')	(Default USD, otherwise can be provided e.g. USD, AUD etc.)						
	->setDescription('New description')			
	->setItemName('New Items for shopping')	//required		
	->setShippingDetails($shipping)	//required only if shipping details are set as above, otherwise optional		
	->setCard($card)	//required only if payment method is set as `credit_card` otherwise not required
	->setIntent('sale')	(Optional)(Default : sale)(e.g. sale, order, authorize)					
	->setDetails($details)						
	->pay();	//required


	b). For the invoice or payment details of any Payment through PayPal


	$payment_id = 'PAY-1JK339012V705711ALPC3QSI';	: Required(fetch this id, when you are making transaction and store it safe)
	$payment_details = \Payment::invoice($payment_id);
	

	c). For refund of any transaction through PayPal
	
	$transaction_id = '3S7574554G079694D'; : Required(fetch this id, when you are making transaction and store)
	$refund = \Payment::setAmount(0.11)
	->setCurrency('USD')
	->setReason('to refund the amount reason')
	->refund($transaction_id);  
	


2). For Stripe (if `PAYMENT_TYPE` == stripe in .env)
	a). For payment through Stripe
	
	//initialising the card object
	$card = Payment::card();
	$card->setNumber(4242424242424242)	: integer (Required)
	->setExpireYear(2020)	: integer (Required)
	->setExpireMonth(06)	: integer (Required)
	->setCvv(314);	: integer (Required)
		
	//making payments
	$pay = \Payment::setAmount(1)	: integer (Required)
	->setDescription('for whatever')	: string  (Optional)
	->setCard($card)	: object  (Required)
	->pay();	: required
	

	b). To get invoice or payment details

	$payment_id = 'ch_1DK0jWB108n83JtHxlqSCqrK';	: Required(fetch this id, when you are making transaction and store it safe)
	
	$payment_details = \Payment::invoice($payment_id);
	

	c). To get refund of any transaction
	
	$transaction_id = 'ch_1DK0j1B108n83JtH236d8aH5'; : Required(fetch this id, when you are making transaction and store)
	$refund = \Payment::setAmount(1)
	->setReason('duplicate')
	->refund($transaction_id); 
	


## Testing

``` bash
$ vendor hestalabs/payments
```

## Environment

a). Just change the environment from `env` for PayPal *sandbox* for testing and *live* for production, in your .env 

```
PAYPAL_MODE = sandbox/live
```

b). For Stripe just change the credentials of live and after activation of account for live payments everything will be fine.


## Security

If you discover any security related issues, please email friends@hestabit.com instead of using the issue tracker.


## License

license. Please see the [license file](license.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/hestalabs/payments.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/hestalabs/payments.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/hestalabs/payments/master.svg?style=flat-square
[ico-styleci]: https://styleci.io/repos/12345678/shield

[link-packagist]: https://packagist.org/packages/hestalabs/payments
[link-downloads]: https://packagist.org/packages/hestalabs/payments
[link-travis]: https://travis-ci.org/hestalabs/payments
[link-author]: https://github.com/hestalabs
[link-contributors]: ../../contributors]

changelog.md
# Changelog

All notable changes to `Payments` will be documented in this file.

## Version 1.0

### Added
- Everything


src/Repositories/PayPal.php
<?php

namespace Hestalabs\Payments\Repositories;

use Validator;	
use PayPal\Api\Sale;
use PayPal\Api\Item;
use PayPal\Api\Payer;
use PayPal\Api\Refund;
use PayPal\Api\Amount;
use PayPal\Api\Payment;
use PayPal\Api\ItemList;
use PayPal\Api\CreditCard;
use PayPal\Api\Transaction;
use PayPal\Rest\ApiContext;
use PayPal\Api\RedirectUrls;
use Illuminate\Http\Request;
use Hestalabs\Payments\Handlers\Setter;
use Anouar\Paypalpayment\PaypalPayment;
use Hestalabs\Payments\Payment as PPPayment;

/*
|--------------------------------------------------
| Handling all the payment behaviours via PayPal.
|--------------------------------------------------
| Written By- Hestalabs
*/
class PayPal extends Setter implements PPPayment{
	/**
	 * Initialising the class with PatPal credentials
	 */
	public function __construct(){
		$config = [
            'mode' => config('payments.paypal_payment.mode'),

            'account' => [
                'client_id' => config('payments.paypal_payment.account.client_id'),
                'client_secret' => config('payments.paypal_payment.account.client_secret'),
            ],

            'http' => [
                'connection_time_out' => config('payments.paypal_payment.http.connection_time_out'),
            ],

            'log' => [
                'log_enabled' => config('payments.paypal_payment.log.log_enabled'),
                'file_name' => config('payments.paypal_payment.log.file_name'),
                'log_level' => config('payments.paypal_payment.log.log_level'),
            ],
        ];

		//setup PayPal api context
		$this->paypal = new Paypalpayment($config);

		$this->_api_context = $this->paypal->apiContext(config('payments.paypal_payment.account.client_id'), config('payments.paypal_payment.account.client_secret'));
	}

	/*
	|------------------------------------------------------
	| For paying via paypal(PayPal Express)
	| Written By- Hestalabs
	|------------------------------------------------------
	| @var line1 			: apha_numeric
	| @var line2 			: apha_numeric
	| @var city 			: string
	| @var state			: string
	| @var postal_code		: integer
	| @var country_code 	: integer
	| @var phone 			: integer
	| @var recipient_name	: string
	| @var type				: string
	| @var number			: integer
	| @var expire_month 	: integer
	| @var expire_year 		: integer
	| @var cvv				: integer
	| @var first_name		: string
	| @var last_name		: string
	| @var shipping_charge 	: numeric
	| @var shipping_tax 	: numeric
	| @var subtotal			: numeric
	| @var tax				: numeric
	| @var price			: numeric
	| @var quantity			: integer
	| @var currency 		: string  (Default : USD)
	| @var item_name 		: string
	| @var return_url		: string
	| @var cancel_url		: string
	| @var intent 			: string  (Default : sale)
	| @var description		: string  (Default : No description avaliable.)
	|--------------------------------------------------------------------------
	| Below are the example how you can set the objects	|
	| //for shipping details 					: (All are optional fields in shipping)
    | $shipping = \Payment::shipping();
    | $shipping->setLine1("3909 Witmer Road")
    | ->setLine2("Niagara Falls")
    | ->setCity("Niagara Falls")
    | ->setState("NY")
    | ->setPostalCode("14305")
    | ->setCountryCode("US")
    | ->setPhone("716-298-1822")
    | ->setRecipientName("Jhone");
	|	
	| //if user wants to add some extra details  	: (All fields are optional in details)
	| $details = \Payment::details();
    | $details//>setShippingCharge(1.0)				: numeric (Optional)
    | ->setShippingTax(1.0)							: numeric (Optional)
    | ->setSubtotal(2.0);							: numeric (Optional)
	|
	| //making payment
	| $pay = \Payment::setTax(0.2)					: numeric (Optional)
    | ->setPrice(1)                           		: numeric (Required)
    | ->setQuantity(1)								: numeric (Optional)(Default : 1)
    | ->setSubTotal(1)								: string  (Optional, If no tax provided)(Default : equals to total)
    | ->setTotal(1.0)                           	: numeric (Required)
    | ->setCurrency('USD')							: string  (Optional)(Default : 'USD')(e.g USD, AUD)
    | ->setDescription('New description')			: string  (Optional)(Default : 'Default description.')
    | ->setItemName('New Items for shopping') 		: string  (Required)
    | ->setShippingDetails($shipping)				: object  (Optional)(If shipping initialised, then Required)
	| ->setReturnUrl(url('/payment?success=true'))	: string  (Optional)(if payment_method is paypal, then Required)(Must be a full and clean URl)
    | ->setCancelUrl(url('/payment?success=false'))	: string  (Optional)(if payment_method is paypal, then Required)(Must be a full and clean URl)
    | ->setExtraParam([		
    |		'url' => true 							: object  (Optional)(If user want only redirection-link of payment, then mark it true)
    |  ])
    |  ->setIntent('sale')							: string  (Optional)(Default : sale)(e.g. sale, order, authorize)
    |  ->setDetails($details)						: object  (Optional)(If details are initialised, then Required)
    |  ->pay();
	*/
	public function pay(){
		try{
			$rules = !empty($this->rules) ? $this->getRules($this->rules) : [];

			$required_arr = [
				'tax'				=> 'numeric|min:0.0',
				'city'				=> 'string',
				'total' 			=> 'required|numeric|min:0.1',
				'line1'				=> 'string',
				'line2'				=> 'string',
				'state'				=> 'string',
				'price'				=> 'required|numeric|min:0.1',	
				'intent'			=> 'string',
				'subtotal'			=> 'numeric|min:0.0',
				'quantity'			=> 'integer|min:1',
				'currency'			=> 'string|min:3',
				'last_name'			=> 'string',
				'item_name'			=> 'required|string',				
				'first_name'		=> 'string',	
				'description'		=> 'string',
				'postal_code'		=> 'integer',		
				'country_code'		=> 'string',
				'shipping_tax'		=> 'numeric|min:0.0',	
				'recipient_name'	=> 'string',		
				'shipping_charge'	=> 'numeric|min:0.0',		
			];

			//if payment via credit card
			if(!empty($this->getCard())){
				return $this->payWithCard();
			}

			$required_arr = [
				'return_url'	=> 'required|url',
				'cancel_url'	=> 'required|url',
			];

			$validator = Validator::make($rules, $required_arr);

			/**
			 * If validator fails
			 */
			if($validator->fails()){
				return dd($validator->messages()->toArray());
			}

	        $payer = $this->paypal
	        ->payer();

	        $payer->setPaymentMethod('paypal');

	        /**
	         * Collecting item's details
	         * @var object
	         */
	        $item = $this->paypal
	        ->item();

	        $item->setName($this->getItemName())
            ->setDescription($this->getDescription())
            ->setCurrency($this->getCurrency())
            ->setQuantity($this->getQuantity())
            ->setTax($this->getTax())
            ->setPrice($this->getPrice());

	        $itemList = $this->paypal
	        ->itemList();

	        $itemList->setItems([ $item ]);
	        
	        //if shipping details are provided
	        if(!empty($this->getShippingDetails())){
	        	$itemList->setShippingAddress($this->getShippingDetails());
	        }

			//Payment Amount
	        $amount = $this->paypal
	        ->amount();

	        $amount->setCurrency($this->getCurrency())
            ->setTotal($this->getTotal());
            
            //if details are provided
            if(!empty($this->getDetails())){
            	$amount->setDetails($this->getDetails());
            }

            /**
             * Transaction details
             * @var object
             */
	        $transaction = $this->paypal->transaction();
	        $transaction->setAmount($amount)
            ->setItemList($itemList)
            ->setDescription($this->getDescription())
            ->setInvoiceNumber($this->getInvoiceNumber());

	        $payment =  $this->paypal->payment();

	        /**
	         * If user have Non-US based account, then the redirect URLs must be given
	         * 
	         */
        	$redirect_urls = new RedirectUrls();

	        $redirect_urls->setReturnUrl($this->getReturnUrl()) //Specify return URL
	        ->setCancelUrl($this->getCancelUrl());				//Specify cancel URL

	        $payment->setPayer($payer)
	        ->setIntent($this->getIntent())            
            ->setTransactions([$transaction]);

            /**
             * In case of paypal-checkout
             * setting the redirect URLs
             */
            $payment->setRedirectUrls($redirect_urls);

	        try {
	        	/**
	        	 * making final payment
	        	 */
	            $payment->create($this->_api_context);
	            
	            /**
	             * if user want only URL of redirectiction
	             * In case of implementation of API
	             * provide $this->extra_param[
	             * 		'url' => true
	             * ]
	             */
	            if($this->extra_param['url']){
	            	/**
	            	 * returns the approval link of payment
	            	 */
	            	return $payment->getApprovalLink();
	            }	            
	        } 
	        catch (\PPConnectionException $ex) {
	            return $ex->getMessage();
	        }

	        /**
	         * redirecting the user to PayPal checkout
	         * @var string
	         */
	        $approval_url = $payment->getApprovalLink();
	        header("Location: {$approval_url}");
	        
	        exit();
		}
		catch(\Exception $e){
			return $e;
		}
	}

	/*
	|------------------------------------------------------
	| For paying via credit_card
	| Written By- Hestalabs
	|------------------------------------------------------
	| @var line1 			: apha_numeric
	| @var line2 			: apha_numeric
	| @var city 			: string
	| @var state			: string
	| @var postal_code		: integer
	| @var country_code 	: integer
	| @var phone 			: integer
	| @var recipient_name	: string
	| @var type				: string
	| @var number			: integer
	| @var expire_month 	: integer
	| @var expire_year 		: integer
	| @var cvv				: integer
	| @var first_name		: string
	| @var last_name		: string
	| @var shipping_charge 	: numeric
	| @var shipping_tax 	: numeric
	| @var subtotal			: numeric
	| @var tax				: numeric
	| @var price			: numeric
	| @var quantity			: integer
	| @var currency 		: string  (Default : USD)
	| @var item_name 		: string
	| @var intent 			: string  (Default : sale)
	| @var description		: string  (Default : No description avaliable.)
	|--------------------------------------------------------------------------
	| Below are the example how you can set the objects	|
	| //for shipping details 					: (All are optional fields in shipping)
    | $shipping = \Payment::shipping();
    | $shipping->setLine1("3909 Witmer Road")
    | ->setLine2("Niagara Falls")
    | ->setCity("Niagara Falls")
    | ->setState("NY")
    | ->setPostalCode("14305")
    | ->setCountryCode("US")
    | ->setPhone("716-298-1822")
    | ->setRecipientName("Jhone");
	|	
	| //if making payment via credit_card 			: (Required, when you select payment_method as credit_card)
	| $card = \Payment::card();
    | $card->setType('visa')						: string  (Optional)
    | ->setNumber(4032038634486363)					: longInt (Required)
    | ->setExpireMonth(11)							: integer (Required)
    | ->setExpireYear(2023)							: integer (Required)
    | ->setCvv(123)									: integer (Required)
    | ->setFirstName('Abhishek')					: string  (Optional)
    | ->setLastName('Rana');						: string  (Optional)
	|	
	| //if user wants to add some extra details  	: (All fields are optional in details)
	| $details = \Payment::details();
    | $details//>setShippingCharge(1.0)				: numeric (Optional)
    | ->setShippingTax(1.0)							: numeric (Optional)
    | ->setSubtotal(2.0);							: numeric (Optional)
	|
	| //making payment
	| $pay = \Payment::setTax(0.2)					: numeric (Optional)
	| ->setPrice(1)									: numeric (Required)
    | ->setAmount(1)                           		: numeric (Required)
    | ->setQuantity(1)								: numeric (Optional)(Default : 1)
    | ->setSubTotal(1)								: string  (Optional, If no tax provided)(Default : equals to total)
    | ->setCurrency('USD')							: string  (Optional)(Default : 'USD')(e.g USD, AUD)
    | ->setDescription('New description')			: string  (Optional)(Default : 'Default description.')
    | ->setItemName('New Items for shopping') 		: string  (Required)
    | ->setShippingDetails($shipping)				: object  (Optional)(If shipping initialised, then Required)
    | ->setCard($card)								: object  (Optional)(If payment_method is credit_card, then Required)
 	| ->setIntent('sale')							: string  (Optional)(Default : sale)(e.g. sale, order, authorize)
    | ->setDetails($details)						: object  (Optional)(If details are initialised, then Required)
    | ->pay();
	*/
	public function payWithCard(){
		try{
			$rules = !empty($this->rules) ? $this->getRules($this->rules) : [];

			$required_arr = [
				'tax'				=> 'numeric|min:0.0',
				'cvv2' 				=> 'required|integer|digits:3',
				'city'				=> 'string',
				'line1'				=> 'string',
				'line2'				=> 'string',
				'state'				=> 'string',
				'price'				=> 'required|numeric|min:0.1',	
				'intent'			=> 'string',
				'amount' 			=> 'required|numeric',
				'number' 			=> 'required|integer|min:1',
				'subtotal'			=> 'numeric|min:0.0',
				'quantity'			=> 'integer|min:1',
				'currency'			=> 'string|min:3',
				'last_name'			=> 'string',
				'item_name'			=> 'required|string',			
				'first_name'		=> 'string',	
				'description'		=> 'string',
				'postal_code'		=> 'integer',		
				'expire_year'		=> 'required|integer|min:1|digits:4',
				'expire_month' 		=> 'required|numeric|min:1|max:12|digits_between:1, 2',
				'country_code'		=> 'string',
				'shipping_tax'		=> 'numeric|min:0.0',	
				'recipient_name'	=> 'string',		
				'shipping_charge'	=> 'numeric|min:0.0',		
			];
			
			$validator = Validator::make($rules, $required_arr , [
				'number.min'		=> 'Card number must be postive.',
				'expire_year.min'	=> 'Expire year must be postive.',
				'expire_month.min'	=> 'Expire month must be postive.',
				'expire_month.digits_between' => 'Expire month must be two digits.',
			]);

			/**
			 * If validator fails
			 */
			if($validator->fails()){
				return dd($validator->messages()->toArray());
			}

			$payer = $this->paypal
	        ->payer();

	        $payer->setPaymentMethod('credit_card');

	        $funding_instrument = $this->paypal
		    ->fundingInstrument();

		    $funding_instrument->setCreditCard($this->getCard());	//setting card details
			$payer->setFundingInstruments([$funding_instrument]);

			$item = $this->paypal
	        ->item();

	        $item->setName($this->getItemName())
            ->setDescription($this->getDescription())
            ->setCurrency($this->getCurrency())
            ->setQuantity($this->getQuantity())
            ->setTax($this->getTax())
            ->setPrice($this->getPrice());

	        $itemList = $this->paypal
	        ->itemList();

	        $itemList->setItems([ $item ]);
	        
	        //if shipping details are provided
	        if(!empty($this->getShippingDetails())){
	        	$itemList->setShippingAddress($this->getShippingDetails());
	        }

	        //Payment Amount
	        $amount = $this->paypal
	        ->amount();

	        $amount->setCurrency($this->getCurrency())
            ->setTotal($this->getAmount());
            
            //if details are provided
            if(!empty($this->getDetails())){
            	$amount->setDetails($this->getDetails());
            }

            /**
             * Transaction details
             * @var object
             */
	        $transaction = $this->paypal->transaction();//Paypalpayment::transaction();
	        $transaction->setAmount($amount)
            ->setItemList($itemList)
            ->setDescription($this->getDescription())
            ->setInvoiceNumber($this->getInvoiceNumber());

	        $payment =  $this->paypal->payment();
	        $payment->setPayer($payer)
	        ->setIntent($this->getIntent())            
            ->setTransactions([$transaction]);

            try {
	        	/**
	        	 * making final payment
	        	 */
	            $payment->create($this->_api_context);	          
	           
	            return $payment;	
	        } 
	        catch (\PPConnectionException $ex) {
	            return $ex->getMessage();
	        }
		}
		catch(\Exception $e){
			return $e;
		}
	}

	/*
	|--------------------------------------------------------------------------
	| Getting the details of given transaction
	| Written By- Hestalabs
	|--------------------------------------------------------------------------
	| @var payment_id 	: integer
	|--------------------------------------------------------------------------
	| Below are the example how you can set the objects
	|
	| $payment_id = 'PAY-1JK339012V705711ALPC3QSI';  	: Required(fetch this id, when you are making transaction and store it safe)	
    | $payment_details = \Payment::invoice($payment_id);
	*/
	public function invoice($payment_id){
		try{			
			return  $this->paypal
			->getById($payment_id, $this->_api_context);	
		}
		catch(\Exception $e){
			return $e;
		}
	}

    /*
	|--------------------------------------------------------
	| To refund the paid amount
	| Written By- Hestalabs
	|--------------------------------------------------------
	| @var reason 			: string 
	| @var amount 			: numeric
	| @var transaction_id	: integer
	|--------------------------------------------------------
	| Below are the example how you can set the objects
	|
	| $transaction_id = '3S7574554G079694D'; : Required(fetch this id, when you are making transaction and store)
	| $refund = \Payment::setAmount(0.11)
    | ->setCurrency('USD')
    | ->setReason('to refund the amount reason')
    | ->refund($transaction_id);  
	*/
	public function refund($transaction_id){
		try{
	        $amt = new Amount();
			$amt->setTotal($this->getAmount()) 		//in paypal its 'total', but we are using 'amount' for the sake of simplicity
			->setCurrency($this->getCurrency());

			$refund = new Refund();
			$refund->setAmount($amt)
			->setReason($this->getReason());

			$sale = new Sale();
			$sale->setId($transaction_id);

			return $sale->refund($refund, $this->_api_context);
		}
		catch(\Exception $e){
			return $e;
		}
	}
}